pennylane/devices/qubit/apply_operation.py

```diff
@@ -260,17 +260,23 @@ def apply_mid_measure(
     """
     if is_state_batched:
         raise ValueError("MidMeasureMP cannot be applied to batched states.")
+    if not np.allclose(np.linalg.norm(state), 1.0):
```

Comment:
why does this happen here, at the beginning of `apply_mid_measure`? I would have expected a check like this after some measurements are applied

Reply:
but because this is at the beginning of the function, doesn't that mean we'll only avoid that issue if someone postselects and makes the state disappear, and then they do _another_ `qml.measure`? which means that if they don't do another, we won't get the benefit of exiting early

Reply:
You are right, this is a very minor optimization since all other operations still get applied until the end of the tape. 

In the deferred measurement implementation, I believe non-normalizable states are just forcefully normalized (to NaNs) and then the rest of the circuit is applied on NaNs, yielding NaNs which are discarded at the end. If we're ok with that, I would return a NaN state instead of zeros and exit from `get_final_state` early as soon as a NaN is detected in `state`. How about that?

Reply:
I would actually be in favour of exiting `get_final_state` early when postselection doesn't match the sample, because that could have a massive speed up in circuits where the probability of the postselected state is low and there are many gates after postselection. @trbromley do you have thoughts here?

Reply:
Trying to exit `get_final_state` prematurely breaks a bunch of tests because autodiff has trouble tracking the NaNs or the `norm` function. I'd suggest we just leave it at that for now.